// ngx-ui-loader/bundles/ngx-ui-loader.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs'), require('@angular/common'), require('@angular/platform-browser'), require('rxjs/operators'), require('@angular/router'), require('@angular/common/http')) :
    typeof define === 'function' && define.amd ? define('ngx-ui-loader', ['exports', '@angular/core', 'rxjs', '@angular/common', '@angular/platform-browser', 'rxjs/operators', '@angular/router', '@angular/common/http'], factory) :
    (global = global || self, factory(global['ngx-ui-loader'] = {}, global.ng.core, global.rxjs, global.ng.common, global.ng.platformBrowser, global.rxjs.operators, global.ng.router, global.ng.common.http));
}(this, (function (exports, core, rxjs, common, platformBrowser, operators, router, http) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __exportStar(m, exports) {
        for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    /**
     * Available spinner types
     */

    (function (SPINNER) {
        SPINNER["ballScaleMultiple"] = "ball-scale-multiple";
        SPINNER["ballSpin"] = "ball-spin";
        SPINNER["ballSpinClockwise"] = "ball-spin-clockwise";
        SPINNER["ballSpinClockwiseFadeRotating"] = "ball-spin-clockwise-fade-rotating";
        SPINNER["ballSpinFadeRotating"] = "ball-spin-fade-rotating";
        SPINNER["chasingDots"] = "chasing-dots";
        SPINNER["circle"] = "circle";
        SPINNER["cubeGrid"] = "cube-grid";
        SPINNER["doubleBounce"] = "double-bounce";
        SPINNER["fadingCircle"] = "fading-circle";
        SPINNER["foldingCube"] = "folding-cube";
        SPINNER["pulse"] = "pulse";
        SPINNER["rectangleBounce"] = "rectangle-bounce";
        SPINNER["rectangleBounceParty"] = "rectangle-bounce-party";
        SPINNER["rectangleBouncePulseOut"] = "rectangle-bounce-pulse-out";
        SPINNER["rectangleBouncePulseOutRapid"] = "rectangle-bounce-pulse-out-rapid";
        SPINNER["rotatingPlane"] = "rotating-plane";
        SPINNER["squareJellyBox"] = "square-jelly-box";
        SPINNER["squareLoader"] = "square-loader";
        SPINNER["threeBounce"] = "three-bounce";
        SPINNER["threeStrings"] = "three-strings";
        SPINNER["wanderingCubes"] = "wandering-cubes";
    })(exports.SPINNER || (exports.SPINNER = {}));
    /**
     * Available postions
     */

    (function (POSITION) {
        POSITION["bottomCenter"] = "bottom-center";
        POSITION["bottomLeft"] = "bottom-left";
        POSITION["bottomRight"] = "bottom-right";
        POSITION["centerCenter"] = "center-center";
        POSITION["centerLeft"] = "center-left";
        POSITION["centerRight"] = "center-right";
        POSITION["topCenter"] = "top-center";
        POSITION["topLeft"] = "top-left";
        POSITION["topRight"] = "top-right";
    })(exports.POSITION || (exports.POSITION = {}));
    /**
     * Progress bar directions
     */

    (function (PB_DIRECTION) {
        PB_DIRECTION["leftToRight"] = "ltr";
        PB_DIRECTION["rightToLeft"] = "rtl";
    })(exports.PB_DIRECTION || (exports.PB_DIRECTION = {}));

    /**
     * The default value of foreground task id
     */
    var DEFAULT_FG_TASK_ID = 'fg-default';
    /**
     * The default value of background task id
     */
    var DEFAULT_BG_TASK_ID = 'bg-default';
    /**
     * The default value of loader id
     */
    var DEFAULT_MASTER_LOADER_ID = 'master';
    var DEFAULT_TIME = {};
    var MIN_DELAY = 0;
    var MIN_TIME = 0;
    var CLOSING_TIME = 1001;
    var FAST_CLOSING_TIME = 601;
    var BACKGROUND = false;
    var FOREGROUND = true;
    var OVERLAY_DISAPPEAR_TIME = 500;
    var FAST_OVERLAY_DISAPPEAR_TIME = 300;
    /**
     * Http loader taskId
     */
    var HTTP_LOADER_TASK_ID = '$_http-loader';
    /**
     * Router loader taskId
     */
    var ROUTER_LOADER_TASK_ID = '$_router_loader';
    /**
     * The configuration of spinners
     */
    var SPINNER_CONFIG = {
        'ball-scale-multiple': {
            divs: 3,
            class: 'sk-ball-scale-multiple'
        },
        'ball-spin': {
            divs: 8,
            class: 'sk-ball-spin'
        },
        'ball-spin-clockwise': {
            divs: 8,
            class: 'sk-ball-spin-clockwise'
        },
        'ball-spin-clockwise-fade-rotating': {
            divs: 8,
            class: 'sk-ball-spin-clockwise-fade-rotating'
        },
        'ball-spin-fade-rotating': {
            divs: 8,
            class: 'sk-ball-spin-fade-rotating'
        },
        'chasing-dots': {
            divs: 2,
            class: 'sk-chasing-dots'
        },
        circle: {
            divs: 12,
            class: 'sk-circle'
        },
        'cube-grid': {
            divs: 9,
            class: 'sk-cube-grid'
        },
        'double-bounce': {
            divs: 2,
            class: 'sk-double-bounce'
        },
        'fading-circle': {
            divs: 12,
            class: 'sk-fading-circle'
        },
        'folding-cube': {
            divs: 4,
            class: 'sk-folding-cube'
        },
        pulse: {
            divs: 1,
            class: 'sk-pulse'
        },
        'rectangle-bounce': {
            divs: 5,
            class: 'sk-rectangle-bounce'
        },
        'rectangle-bounce-party': {
            divs: 5,
            class: 'sk-rectangle-bounce-party'
        },
        'rectangle-bounce-pulse-out': {
            divs: 5,
            class: 'sk-rectangle-bounce-pulse-out'
        },
        'rectangle-bounce-pulse-out-rapid': {
            divs: 5,
            class: 'sk-rectangle-bounce-pulse-out-rapid'
        },
        'rotating-plane': {
            divs: 1,
            class: 'sk-rotating-plane'
        },
        'square-jelly-box': {
            divs: 2,
            class: 'sk-square-jelly-box'
        },
        'square-loader': {
            divs: 1,
            class: 'sk-square-loader'
        },
        'three-bounce': {
            divs: 3,
            class: 'sk-three-bounce'
        },
        'three-strings': {
            divs: 3,
            class: 'sk-three-strings'
        },
        'wandering-cubes': {
            divs: 2,
            class: 'sk-wandering-cubes'
        }
    };
    /**
     * The default configuration of ngx-ui-loader
     */
    var DEFAULT_CONFIG = {
        bgsColor: '#00ACC1',
        bgsOpacity: 0.5,
        bgsPosition: exports.POSITION.bottomRight,
        bgsSize: 60,
        bgsType: exports.SPINNER.ballSpinClockwise,
        blur: 5,
        delay: 0,
        fastFadeOut: false,
        fgsColor: '#00ACC1',
        fgsPosition: exports.POSITION.centerCenter,
        fgsSize: 60,
        fgsType: exports.SPINNER.ballSpinClockwise,
        gap: 24,
        logoPosition: exports.POSITION.centerCenter,
        logoSize: 120,
        logoUrl: '',
        masterLoaderId: DEFAULT_MASTER_LOADER_ID,
        overlayBorderRadius: '0',
        overlayColor: 'rgba(40, 40, 40, 0.8)',
        pbColor: '#00ACC1',
        pbDirection: exports.PB_DIRECTION.leftToRight,
        pbThickness: 3,
        hasProgressBar: true,
        text: '',
        textColor: '#FFFFFF',
        textPosition: exports.POSITION.centerCenter,
        maxTime: -1,
        minTime: 300
    };

    /**
     * Injection token for ngx-ui-loader configuration
     */
    var NGX_UI_LOADER_CONFIG_TOKEN = new core.InjectionToken('ngxUiLoaderCustom.config');

    var NgxUiLoaderService = /** @class */ (function () {
        /**
         * Constructor
         */
        function NgxUiLoaderService(config) {
            this.config = config;
            this.defaultConfig = __assign({}, DEFAULT_CONFIG);
            if (this.config) {
                if (this.config.minTime && this.config.minTime < MIN_TIME) {
                    this.config.minTime = MIN_TIME;
                }
                this.defaultConfig = __assign(__assign({}, this.defaultConfig), this.config);
            }
            this.loaders = {};
            this.showForeground = new rxjs.BehaviorSubject({ loaderId: '', isShow: false });
            this.showForeground$ = this.showForeground.asObservable();
            this.showBackground = new rxjs.BehaviorSubject({ loaderId: '', isShow: false });
            this.showBackground$ = this.showBackground.asObservable();
            this.fgClosing = new rxjs.BehaviorSubject({ loaderId: '', isShow: false });
            this.foregroundClosing$ = this.fgClosing.asObservable();
            this.bgClosing = new rxjs.BehaviorSubject({ loaderId: '', isShow: false });
            this.backgroundClosing$ = this.bgClosing.asObservable();
        }
        /**
         * For internal use only.
         * @docs-private
         */
        NgxUiLoaderService.prototype.bindLoaderData = function (loaderId) {
            var isMaster = loaderId === this.defaultConfig.masterLoaderId;
            if (this.loaders[loaderId]) {
                if (this.loaders[loaderId].isBound) {
                    throw new Error("[ngx-ui-loader] - loaderId \"" + loaderId + "\" is duplicated.");
                }
                this.loaders[loaderId].isBound = true;
                this.loaders[loaderId].isMaster = isMaster;
                // emit showEvent after data loader is bound
                if (this.hasRunningTask(FOREGROUND, loaderId)) {
                    this.showForeground.next({ loaderId: loaderId, isShow: true });
                }
                else {
                    if (this.hasRunningTask(BACKGROUND, loaderId)) {
                        this.showBackground.next({ loaderId: loaderId, isShow: true });
                    }
                }
            }
            else {
                this.createLoaderData(loaderId, isMaster, true);
            }
        };
        /**
         * For internal use only.
         * @docs-private
         */
        NgxUiLoaderService.prototype.destroyLoaderData = function (loaderId) {
            this.stopAllLoader(loaderId);
            delete this.loaders[loaderId];
        };
        /**
         * Get default loader configuration
         * @returns default configuration object
         */
        NgxUiLoaderService.prototype.getDefaultConfig = function () {
            return __assign({}, this.defaultConfig);
        };
        /**
         * Get all the loaders
         */
        NgxUiLoaderService.prototype.getLoaders = function () {
            return JSON.parse(JSON.stringify(this.loaders));
        };
        /**
         * Get data of a specified loader. If loaderId is not provided, it will return data of
         * master loader(if existed). Otherwise null is returned.
         */
        NgxUiLoaderService.prototype.getLoader = function (loaderId) {
            loaderId = loaderId ? loaderId : this.defaultConfig.masterLoaderId;
            if (this.loaders[loaderId]) {
                return JSON.parse(JSON.stringify(this.loaders[loaderId]));
            }
            return null;
        };
        /**
         * Start the foreground loading of loader having `loaderId` with a specified `taskId`.
         * The loading is only closed off when all taskIds of that loader are called with stopLoader() method.
         * @param loaderId the loader Id
         * @param taskId the optional task Id of the loading. taskId is set to 'fd-default' by default.
         */
        NgxUiLoaderService.prototype.startLoader = function (loaderId, taskId, time) {
            if (taskId === void 0) { taskId = DEFAULT_FG_TASK_ID; }
            if (!this.readyToStart(loaderId, taskId, true, time)) {
                return;
            }
            if (!this.loaders[loaderId].tasks[taskId].isOtherRunning) {
                // no other foreground task running
                if (this.hasRunningTask(BACKGROUND, loaderId)) {
                    this.backgroundCloseout(loaderId);
                    this.showBackground.next({ loaderId: loaderId, isShow: false });
                }
                this.showForeground.next({ loaderId: loaderId, isShow: true });
            }
        };
        /**
         * Start the foreground loading of master loader with a specified `taskId`.
         * The loading is only closed off when all taskIds of that loader are called with stop() method.
         * NOTE: Really this function just wraps startLoader() function
         * @param taskId the optional task Id of the loading. taskId is set to 'fd-default' by default.
         */
        NgxUiLoaderService.prototype.start = function (taskId, time) {
            if (taskId === void 0) { taskId = DEFAULT_FG_TASK_ID; }
            this.startLoader(this.defaultConfig.masterLoaderId, taskId, time);
        };
        /**
         * Start the background loading of loader having `loaderId` with a specified `taskId`.
         * The loading is only closed off when all taskIds of that loader are called with stopLoaderBackground() method.
         * @param loaderId the loader Id
         * @param taskId the optional task Id of the loading. taskId is set to 'bg-default' by default.
         */
        NgxUiLoaderService.prototype.startBackgroundLoader = function (loaderId, taskId, time) {
            if (taskId === void 0) { taskId = DEFAULT_BG_TASK_ID; }
            if (!this.readyToStart(loaderId, taskId, false, time)) {
                return;
            }
            if (!this.hasRunningTask(FOREGROUND, loaderId) && !this.loaders[loaderId].tasks[taskId].isOtherRunning) {
                this.showBackground.next({ loaderId: loaderId, isShow: true });
            }
        };
        /**
         * Start the background loading of master loader with a specified `taskId`.
         * The loading is only closed off when all taskIds of that loader are called with stopBackground() method.
         * NOTE: Really this function just wraps startBackgroundLoader() function
         * @param taskId the optional task Id of the loading. taskId is set to 'bg-default' by default.
         */
        NgxUiLoaderService.prototype.startBackground = function (taskId, time) {
            if (taskId === void 0) { taskId = DEFAULT_BG_TASK_ID; }
            this.startBackgroundLoader(this.defaultConfig.masterLoaderId, taskId, time);
        };
        /**
         * Stop a foreground loading of loader having `loaderId` with specific `taskId`
         * @param loaderId the loader Id
         * @param taskId the optional task Id to stop. If not provided, 'fg-default' is used.
         * @returns Object
         */
        NgxUiLoaderService.prototype.stopLoader = function (loaderId, taskId) {
            var _this = this;
            if (taskId === void 0) { taskId = DEFAULT_FG_TASK_ID; }
            if (!this.readyToStop(loaderId, taskId)) {
                return;
            }
            if (!this.hasRunningTask(FOREGROUND, loaderId)) {
                this.foregroundCloseout(loaderId);
                this.showForeground.next({ loaderId: loaderId, isShow: false });
                if (this.hasRunningTask(BACKGROUND, loaderId)) {
                    setTimeout(function () {
                        if (_this.hasRunningTask(BACKGROUND, loaderId)) {
                            // still have background tasks
                            _this.showBackground.next({ loaderId: loaderId, isShow: true });
                        }
                    }, this.defaultConfig.fastFadeOut ? FAST_OVERLAY_DISAPPEAR_TIME : OVERLAY_DISAPPEAR_TIME);
                }
            }
        };
        /**
         * Stop a foreground loading of master loader with specific `taskId`
         * @param taskId the optional task Id to stop. If not provided, 'fg-default' is used.
         * @returns Object
         */
        NgxUiLoaderService.prototype.stop = function (taskId) {
            if (taskId === void 0) { taskId = DEFAULT_FG_TASK_ID; }
            this.stopLoader(this.defaultConfig.masterLoaderId, taskId);
        };
        /**
         * Stop a background loading of loader having `loaderId` with specific `taskId`
         * @param loaderId the loader Id
         * @param taskId the optional task Id to stop. If not provided, 'bg-default' is used.
         * @returns Object
         */
        NgxUiLoaderService.prototype.stopBackgroundLoader = function (loaderId, taskId) {
            if (taskId === void 0) { taskId = DEFAULT_BG_TASK_ID; }
            if (!this.readyToStop(loaderId, taskId)) {
                return;
            }
            if (!this.hasRunningTask(FOREGROUND, loaderId) && !this.hasRunningTask(BACKGROUND, loaderId)) {
                this.backgroundCloseout(loaderId);
                this.showBackground.next({ loaderId: loaderId, isShow: false });
            }
        };
        /**
         * Stop a background loading of master loader with specific taskId
         * @param taskId the optional task Id to stop. If not provided, 'bg-default' is used.
         * @returns Object
         */
        NgxUiLoaderService.prototype.stopBackground = function (taskId) {
            if (taskId === void 0) { taskId = DEFAULT_BG_TASK_ID; }
            this.stopBackgroundLoader(this.defaultConfig.masterLoaderId, taskId);
        };
        /**
         * Stop all the background and foreground loadings of loader having `loaderId`
         * @param loaderId the loader Id
         */
        NgxUiLoaderService.prototype.stopAllLoader = function (loaderId) {
            if (!this.loaders[loaderId]) {
                console.warn("[ngx-ui-loader] - loaderId \"" + loaderId + "\" does not exist.");
                return;
            }
            if (this.hasRunningTask(FOREGROUND, loaderId)) {
                this.foregroundCloseout(loaderId);
                this.showForeground.next({ loaderId: loaderId, isShow: false });
            }
            else if (this.hasRunningTask(BACKGROUND, loaderId)) {
                this.backgroundCloseout(loaderId);
                this.showBackground.next({ loaderId: loaderId, isShow: false });
            }
            this.clearAllTimers(this.loaders[loaderId].tasks);
            this.loaders[loaderId].tasks = {};
        };
        /**
         * Stop all the background and foreground loadings of master loader
         */
        NgxUiLoaderService.prototype.stopAll = function () {
            this.stopAllLoader(this.defaultConfig.masterLoaderId);
        };
        /**
         * Create loader data if it does not exist
         * @docs-private
         */
        NgxUiLoaderService.prototype.createLoaderData = function (loaderId, isMaster, isBound) {
            if (!this.loaders[loaderId]) {
                this.loaders[loaderId] = {
                    loaderId: loaderId,
                    tasks: {},
                    isMaster: isMaster,
                    isBound: isBound
                };
            }
        };
        /**
         * Manage to close foreground loading
         * @docs-private
         * @param loaderId the loader id
         */
        NgxUiLoaderService.prototype.foregroundCloseout = function (loaderId) {
            var _this = this;
            this.fgClosing.next({ loaderId: loaderId, isShow: true });
            setTimeout(function () {
                _this.fgClosing.next({ loaderId: loaderId, isShow: false });
            }, this.defaultConfig.fastFadeOut ? FAST_CLOSING_TIME : CLOSING_TIME);
        };
        /**
         * Manage to close background loading
         * @docs-private
         * @param loaderId the loader id
         */
        NgxUiLoaderService.prototype.backgroundCloseout = function (loaderId) {
            var _this = this;
            this.bgClosing.next({ loaderId: loaderId, isShow: true });
            setTimeout(function () {
                _this.bgClosing.next({ loaderId: loaderId, isShow: false });
            }, this.defaultConfig.fastFadeOut ? FAST_CLOSING_TIME : CLOSING_TIME);
        };
        /**
         * Clear all timers of the given task
         * @docs-private
         */
        NgxUiLoaderService.prototype.clearTimers = function (task) {
            clearTimeout(task.delayTimer);
            clearTimeout(task.maxTimer);
            clearTimeout(task.minTimer);
        };
        /**
         * Clear all timers of the given tasks
         * @docs-private
         */
        NgxUiLoaderService.prototype.clearAllTimers = function (tasks) {
            var _this = this;
            Object.keys(tasks).map(function (id) {
                _this.clearTimers(tasks[id]);
            });
        };
        /**
         * Check whether the specified loader has a running task with the given `taskId`.
         * If no `taskId` specified, it will check whether the loader has any running tasks.
         * For internal use only.
         * @docs-private
         * @param isForeground foreground task or background task
         * @param loaderId the loader Id
         * @param taskId the optional task Id
         * @returns boolean
         */
        NgxUiLoaderService.prototype.hasRunningTask = function (isForeground, loaderId, taskId) {
            if (this.loaders[loaderId]) {
                var tasks_1 = this.loaders[loaderId].tasks;
                if (taskId) {
                    return tasks_1[taskId] ? (tasks_1[taskId].startAt ? true : false) : false;
                }
                return Object.keys(tasks_1).some(function (id) { return !!tasks_1[id].startAt && tasks_1[id].isForeground === isForeground; });
            }
            return false;
        };
        /**
         * @docs-private
         */
        NgxUiLoaderService.prototype.readyToStart = function (loaderId, taskId, isForeground, time) {
            if (time === void 0) { time = DEFAULT_TIME; }
            this.createLoaderData(loaderId, undefined, false);
            var isOtherRunning = this.hasRunningTask(isForeground, loaderId);
            if (!this.loaders[loaderId].tasks[taskId]) {
                this.loaders[loaderId].tasks[taskId] = {
                    taskId: taskId,
                    isForeground: isForeground,
                    minTime: time.minTime >= MIN_TIME ? time.minTime : this.defaultConfig.minTime,
                    maxTime: time.maxTime ? time.maxTime : this.defaultConfig.maxTime,
                    delay: time.delay >= MIN_DELAY ? time.delay : this.defaultConfig.delay
                };
            }
            else {
                if (this.loaders[loaderId].tasks[taskId].isForeground !== isForeground) {
                    throw new Error("[ngx-ui-loader] - taskId \"" + taskId + "\" is duplicated.");
                }
            }
            if (this.setDelayTimer(this.loaders[loaderId].tasks[taskId], loaderId)) {
                return false;
            }
            this.loaders[loaderId].tasks[taskId] = __assign(__assign({}, this.loaders[loaderId].tasks[taskId]), { isOtherRunning: isOtherRunning, startAt: Date.now() });
            this.setMaxTimer(this.loaders[loaderId].tasks[taskId], loaderId);
            if (!this.loaders[loaderId].isBound) {
                return false;
            }
            return true;
        };
        /**
         * @docs-private
         */
        NgxUiLoaderService.prototype.readyToStop = function (loaderId, taskId) {
            if (!this.loaders[loaderId]) {
                console.warn("[ngx-ui-loader] - loaderId \"" + loaderId + "\" does not exist.");
                return false;
            }
            var task = this.loaders[loaderId].tasks[taskId];
            if (!task) {
                return false;
            }
            if (task.isDelayed) {
                this.clearTimers(task);
                delete this.loaders[loaderId].tasks[taskId];
                return false;
            }
            if (this.setMinTimer(task, loaderId)) {
                return false;
            }
            this.clearTimers(task);
            delete this.loaders[loaderId].tasks[taskId];
            return true;
        };
        /**
         * Set delay timer, if `delay` > 0
         * @docs-private
         * @returns boolean
         */
        NgxUiLoaderService.prototype.setDelayTimer = function (task, loaderId) {
            var _this = this;
            if (task.delay > MIN_DELAY) {
                if (task.isDelayed) {
                    return true;
                }
                if (!task.delayTimer) {
                    task.isDelayed = true;
                    task.delayTimer = setTimeout(function () {
                        task.isDelayed = false;
                        if (task.isForeground) {
                            _this.startLoader(loaderId, task.taskId);
                        }
                        else {
                            _this.startBackgroundLoader(loaderId, task.taskId);
                        }
                    }, task.delay);
                    return true;
                }
            }
            return false;
        };
        /**
         * Set maxTimer if `maxTime` > `minTime`
         * @docs-private
         * @returns boolean
         */
        NgxUiLoaderService.prototype.setMaxTimer = function (task, loaderId) {
            var _this = this;
            if (task.maxTime > task.minTime) {
                // restart the task, reset maxTimer
                clearTimeout(task.maxTimer);
                task.maxTimer = setTimeout(function () {
                    if (task.isForeground) {
                        _this.stopLoader(loaderId, task.taskId);
                    }
                    else {
                        _this.stopBackgroundLoader(loaderId, task.taskId);
                    }
                }, task.maxTime);
            }
        };
        /**
         * Set minTimer if `startAt` + `minTime` > `Date.now()`
         * @docs-private
         * @returns boolean
         */
        NgxUiLoaderService.prototype.setMinTimer = function (task, loaderId) {
            var _this = this;
            var now = Date.now();
            if (task.startAt) {
                if (task.startAt + task.minTime > now) {
                    task.minTimer = setTimeout(function () {
                        if (task.isForeground) {
                            _this.stopLoader(loaderId, task.taskId);
                        }
                        else {
                            _this.stopBackgroundLoader(loaderId, task.taskId);
                        }
                    }, task.startAt + task.minTime - now);
                    return true;
                }
            }
            return false;
        };
        NgxUiLoaderService.ctorParameters = function () { return [
            { type: undefined, decorators: [{ type: core.Optional }, { type: core.Inject, args: [NGX_UI_LOADER_CONFIG_TOKEN,] }] }
        ]; };
        NgxUiLoaderService.ɵprov = core["ɵɵdefineInjectable"]({ factory: function NgxUiLoaderService_Factory() { return new NgxUiLoaderService(core["ɵɵinject"](NGX_UI_LOADER_CONFIG_TOKEN, 8)); }, token: NgxUiLoaderService, providedIn: "root" });
        NgxUiLoaderService = __decorate([
            core.Injectable({
                providedIn: 'root'
            }),
            __param(0, core.Optional()), __param(0, core.Inject(NGX_UI_LOADER_CONFIG_TOKEN)),
            __metadata("design:paramtypes", [Object])
        ], NgxUiLoaderService);
        return NgxUiLoaderService;
    }());

    var NgxUiLoaderComponent = /** @class */ (function () {
        /**
         * Constructor
         */
        function NgxUiLoaderComponent(domSanitizer, changeDetectorRef, ngxService) {
            this.domSanitizer = domSanitizer;
            this.changeDetectorRef = changeDetectorRef;
            this.ngxService = ngxService;
            this.initialized = false;
            this.defaultConfig = this.ngxService.getDefaultConfig();
            this.bgsColor = this.defaultConfig.bgsColor;
            this.bgsOpacity = this.defaultConfig.bgsOpacity;
            this.bgsPosition = this.defaultConfig.bgsPosition;
            this.bgsSize = this.defaultConfig.bgsSize;
            this.bgsType = this.defaultConfig.bgsType;
            this.fastFadeOut = this.defaultConfig.fastFadeOut;
            this.fgsColor = this.defaultConfig.fgsColor;
            this.fgsPosition = this.defaultConfig.fgsPosition;
            this.fgsSize = this.defaultConfig.fgsSize;
            this.fgsType = this.defaultConfig.fgsType;
            this.gap = this.defaultConfig.gap;
            this.loaderId = this.defaultConfig.masterLoaderId;
            this.logoPosition = this.defaultConfig.logoPosition;
            this.logoSize = this.defaultConfig.logoSize;
            this.logoUrl = this.defaultConfig.logoUrl;
            this.overlayBorderRadius = this.defaultConfig.overlayBorderRadius;
            this.overlayColor = this.defaultConfig.overlayColor;
            this.pbColor = this.defaultConfig.pbColor;
            this.pbDirection = this.defaultConfig.pbDirection;
            this.pbThickness = this.defaultConfig.pbThickness;
            this.hasProgressBar = this.defaultConfig.hasProgressBar;
            this.text = this.defaultConfig.text;
            this.textColor = this.defaultConfig.textColor;
            this.textPosition = this.defaultConfig.textPosition;
        }
        /**
         * On init event
         */
        NgxUiLoaderComponent.prototype.ngOnInit = function () {
            var _this = this;
            this.initializeSpinners();
            this.ngxService.bindLoaderData(this.loaderId);
            this.determinePositions();
            this.trustedLogoUrl = this.domSanitizer.bypassSecurityTrustResourceUrl(this.logoUrl);
            this.showForegroundWatcher = this.ngxService.showForeground$
                .pipe(operators.filter(function (showEvent) { return _this.loaderId === showEvent.loaderId; }))
                .subscribe(function (data) {
                _this.showForeground = data.isShow;
                _this.changeDetectorRef.markForCheck();
            });
            this.showBackgroundWatcher = this.ngxService.showBackground$
                .pipe(operators.filter(function (showEvent) { return _this.loaderId === showEvent.loaderId; }))
                .subscribe(function (data) {
                _this.showBackground = data.isShow;
                _this.changeDetectorRef.markForCheck();
            });
            this.foregroundClosingWatcher = this.ngxService.foregroundClosing$
                .pipe(operators.filter(function (showEvent) { return _this.loaderId === showEvent.loaderId; }))
                .subscribe(function (data) {
                _this.foregroundClosing = data.isShow;
                _this.changeDetectorRef.markForCheck();
            });
            this.backgroundClosingWatcher = this.ngxService.backgroundClosing$
                .pipe(operators.filter(function (showEvent) { return _this.loaderId === showEvent.loaderId; }))
                .subscribe(function (data) {
                _this.backgroundClosing = data.isShow;
                _this.changeDetectorRef.markForCheck();
            });
            this.initialized = true;
        };
        /**
         * On changes event
         */
        NgxUiLoaderComponent.prototype.ngOnChanges = function (changes) {
            if (!this.initialized) {
                return;
            }
            var bgsTypeChange = changes.bgsType;
            var fgsTypeChange = changes.fgsType;
            var logoUrlChange = changes.logoUrl;
            if (fgsTypeChange || bgsTypeChange) {
                this.initializeSpinners();
            }
            this.determinePositions();
            if (logoUrlChange) {
                this.trustedLogoUrl = this.domSanitizer.bypassSecurityTrustResourceUrl(this.logoUrl);
            }
        };
        /**
         * Initialize spinners
         */
        NgxUiLoaderComponent.prototype.initializeSpinners = function () {
            this.fgDivs = Array(SPINNER_CONFIG[this.fgsType].divs).fill(1);
            this.fgSpinnerClass = SPINNER_CONFIG[this.fgsType].class;
            this.bgDivs = Array(SPINNER_CONFIG[this.bgsType].divs).fill(1);
            this.bgSpinnerClass = SPINNER_CONFIG[this.bgsType].class;
        };
        /**
         * Determine the positions of spinner, logo and text
         */
        NgxUiLoaderComponent.prototype.determinePositions = function () {
            this.logoTop = 'initial';
            this.spinnerTop = 'initial';
            this.textTop = 'initial';
            var textSize = 24;
            if (this.logoPosition.startsWith('center')) {
                this.logoTop = '50%';
            }
            else if (this.logoPosition.startsWith('top')) {
                this.logoTop = '30px';
            }
            if (this.fgsPosition.startsWith('center')) {
                this.spinnerTop = '50%';
            }
            else if (this.fgsPosition.startsWith('top')) {
                this.spinnerTop = '30px';
            }
            if (this.textPosition.startsWith('center')) {
                this.textTop = '50%';
            }
            else if (this.textPosition.startsWith('top')) {
                this.textTop = '30px';
            }
            if (this.fgsPosition === exports.POSITION.centerCenter) {
                if (this.logoUrl && this.logoPosition === exports.POSITION.centerCenter) {
                    if (this.text && this.textPosition === exports.POSITION.centerCenter) {
                        // logo, spinner and text
                        this.logoTop = this.domSanitizer.bypassSecurityTrustStyle("calc(50% - " + this.fgsSize / 2 + "px - " + textSize / 2 + "px - " + this.gap + "px)");
                        this.spinnerTop = this.domSanitizer.bypassSecurityTrustStyle("calc(50% + " + this.logoSize / 2 + "px - " + textSize / 2 + "px)");
                        this.textTop = this.domSanitizer.bypassSecurityTrustStyle("calc(50% + " + this.logoSize / 2 + "px + " + this.gap + "px + " + this.fgsSize / 2 + "px)");
                    }
                    else {
                        // logo and spinner
                        this.logoTop = this.domSanitizer.bypassSecurityTrustStyle("calc(50% - " + this.fgsSize / 2 + "px - " + this.gap / 2 + "px)");
                        this.spinnerTop = this.domSanitizer.bypassSecurityTrustStyle("calc(50% + " + this.logoSize / 2 + "px + " + this.gap / 2 + "px)");
                    }
                }
                else {
                    if (this.text && this.textPosition === exports.POSITION.centerCenter) {
                        // spinner and text
                        this.spinnerTop = this.domSanitizer.bypassSecurityTrustStyle("calc(50% - " + textSize / 2 + "px - " + this.gap / 2 + "px)");
                        this.textTop = this.domSanitizer.bypassSecurityTrustStyle("calc(50% + " + this.fgsSize / 2 + "px + " + this.gap / 2 + "px)");
                    }
                }
            }
            else {
                if (this.logoUrl && this.logoPosition === exports.POSITION.centerCenter && this.text && this.textPosition === exports.POSITION.centerCenter) {
                    // logo and text
                    this.logoTop = this.domSanitizer.bypassSecurityTrustStyle("calc(50% - " + textSize / 2 + "px - " + this.gap / 2 + "px)");
                    this.textTop = this.domSanitizer.bypassSecurityTrustStyle("calc(50% + " + this.logoSize / 2 + "px + " + this.gap / 2 + "px)");
                }
            }
        };
        /**
         * On destroy event
         */
        NgxUiLoaderComponent.prototype.ngOnDestroy = function () {
            this.ngxService.destroyLoaderData(this.loaderId);
            if (this.showForegroundWatcher) {
                this.showForegroundWatcher.unsubscribe();
            }
            if (this.showBackgroundWatcher) {
                this.showBackgroundWatcher.unsubscribe();
            }
            if (this.foregroundClosingWatcher) {
                this.foregroundClosingWatcher.unsubscribe();
            }
            if (this.backgroundClosingWatcher) {
                this.backgroundClosingWatcher.unsubscribe();
            }
        };
        NgxUiLoaderComponent.ctorParameters = function () { return [
            { type: platformBrowser.DomSanitizer },
            { type: core.ChangeDetectorRef },
            { type: NgxUiLoaderService }
        ]; };
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxUiLoaderComponent.prototype, "bgsColor", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Number)
        ], NgxUiLoaderComponent.prototype, "bgsOpacity", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxUiLoaderComponent.prototype, "bgsPosition", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Number)
        ], NgxUiLoaderComponent.prototype, "bgsSize", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", core.TemplateRef)
        ], NgxUiLoaderComponent.prototype, "bgsTemplate", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxUiLoaderComponent.prototype, "bgsType", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxUiLoaderComponent.prototype, "fgsColor", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxUiLoaderComponent.prototype, "fgsPosition", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Number)
        ], NgxUiLoaderComponent.prototype, "fgsSize", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", core.TemplateRef)
        ], NgxUiLoaderComponent.prototype, "fgsTemplate", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxUiLoaderComponent.prototype, "fgsType", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Number)
        ], NgxUiLoaderComponent.prototype, "gap", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxUiLoaderComponent.prototype, "loaderId", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxUiLoaderComponent.prototype, "logoPosition", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Number)
        ], NgxUiLoaderComponent.prototype, "logoSize", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxUiLoaderComponent.prototype, "logoUrl", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxUiLoaderComponent.prototype, "overlayBorderRadius", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxUiLoaderComponent.prototype, "overlayColor", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxUiLoaderComponent.prototype, "pbColor", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxUiLoaderComponent.prototype, "pbDirection", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Number)
        ], NgxUiLoaderComponent.prototype, "pbThickness", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Boolean)
        ], NgxUiLoaderComponent.prototype, "hasProgressBar", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxUiLoaderComponent.prototype, "text", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxUiLoaderComponent.prototype, "textColor", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxUiLoaderComponent.prototype, "textPosition", void 0);
        NgxUiLoaderComponent = __decorate([
            core.Component({
                selector: 'ngx-ui-loader',
                template: "<!-- Progress bar {{{ -->\n<div\n  *ngIf=\"hasProgressBar\"\n  class=\"ngx-progress-bar\"\n  [class.ngx-position-absolute]=\"loaderId !== defaultConfig.masterLoaderId\"\n  [ngClass]=\"'ngx-progress-bar-' + pbDirection\"\n  [style.height.px]=\"pbThickness\"\n  [style.color]=\"pbColor\"\n  [class.loading-foreground]=\"showForeground\"\n  [class.foreground-closing]=\"foregroundClosing\"\n  [class.fast-closing]=\"foregroundClosing && fastFadeOut\"\n></div>\n<!-- Progress bar }}} -->\n\n<!-- Foreground container {{{ -->\n<div\n  class=\"ngx-overlay\"\n  [class.ngx-position-absolute]=\"loaderId !== defaultConfig.masterLoaderId\"\n  [style.background-color]=\"overlayColor\"\n  [style.border-radius]=\"overlayBorderRadius\"\n  [class.loading-foreground]=\"showForeground\"\n  [class.foreground-closing]=\"foregroundClosing\"\n  [class.fast-closing]=\"foregroundClosing && fastFadeOut\"\n>\n  <!-- Logo {{{ -->\n  <img\n    *ngIf=\"logoUrl\"\n    class=\"ngx-loading-logo\"\n    [ngClass]=\"logoPosition\"\n    [src]=\"trustedLogoUrl\"\n    [style.width.px]=\"logoSize\"\n    [style.height.px]=\"logoSize\"\n    [style.top]=\"logoTop\"\n  />\n  <!-- Logo }}} -->\n\n  <!-- Foreground spinner {{{ -->\n  <div\n    class=\"ngx-foreground-spinner\"\n    [ngClass]=\"fgsPosition\"\n    [style.color]=\"fgsColor\"\n    [style.width.px]=\"fgsSize\"\n    [style.height.px]=\"fgsSize\"\n    [style.top]=\"spinnerTop\"\n  >\n    <div *ngIf=\"!fgsTemplate; else foregroundTemplate\" [class]=\"fgSpinnerClass\">\n      <div *ngFor=\"let div of fgDivs\"></div>\n    </div>\n    <ng-template #foregroundTemplate>\n      <ng-container *ngTemplateOutlet=\"fgsTemplate\"></ng-container>\n    </ng-template>\n  </div>\n  <!-- Foreground spinner }}} -->\n\n  <!-- Loading text {{{ -->\n  <div class=\"ngx-loading-text\" [ngClass]=\"textPosition\" [style.top]=\"textTop\" [style.color]=\"textColor\">{{ text }}</div>\n  <!-- Loading text }}} -->\n</div>\n<!-- Foreground container }}} -->\n\n<!-- Background spinner {{{ -->\n<div\n  class=\"ngx-background-spinner\"\n  [class.ngx-position-absolute]=\"loaderId !== defaultConfig.masterLoaderId\"\n  [ngClass]=\"bgsPosition\"\n  [class.loading-background]=\"showBackground\"\n  [class.background-closing]=\"backgroundClosing\"\n  [class.fast-closing]=\"backgroundClosing && fastFadeOut\"\n  [style.width.px]=\"bgsSize\"\n  [style.height.px]=\"bgsSize\"\n  [style.color]=\"bgsColor\"\n  [style.opacity]=\"bgsOpacity\"\n>\n  <div *ngIf=\"!bgsTemplate; else backgroundTemplate\" [class]=\"bgSpinnerClass\">\n    <div *ngFor=\"let div of bgDivs\"></div>\n  </div>\n  <ng-template #backgroundTemplate>\n    <ng-container *ngTemplateOutlet=\"bgsTemplate\"></ng-container>\n  </ng-template>\n</div>\n<!-- Background spinner }}} -->\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                styles: [".ngx-progress-bar{position:fixed;top:0;left:0;width:100%;height:3px;z-index:99999!important;display:none;color:#00acc1;overflow:hidden}.ngx-progress-bar.foreground-closing,.ngx-progress-bar.loading-foreground{display:block}.ngx-progress-bar.foreground-closing{opacity:0!important;transition:opacity .5s ease-out .5s}.ngx-progress-bar.fast-closing{transition:opacity .3s ease-out .3s!important}.ngx-progress-bar::after,.ngx-progress-bar::before{background-color:currentColor;content:\"\";display:block;width:100%;height:100%;position:absolute;top:0}.ngx-progress-bar-ltr::before{transform:translate3d(-100%,0,0)}.ngx-progress-bar-ltr::after{-webkit-animation:12s ease-out progressBar-slide-ltr;animation:12s ease-out progressBar-slide-ltr;transform:translate3d(-5%,0,0)}.ngx-progress-bar-rtl::before{transform:translate3d(100%,0,0)}.ngx-progress-bar-rtl::after{-webkit-animation:12s ease-out progressBar-slide-rtl;animation:12s ease-out progressBar-slide-rtl;transform:translate3d(5%,0,0)}.foreground-closing.ngx-progress-bar-ltr::before{-webkit-animation:1s ease-out progressBar-slide-complete-ltr;animation:1s ease-out progressBar-slide-complete-ltr;transform:translate3d(0,0,0)}.fast-closing.ngx-progress-bar-ltr::before{-webkit-animation:.6s ease-out progressBar-slide-complete-ltr!important;animation:.6s ease-out progressBar-slide-complete-ltr!important}.foreground-closing.ngx-progress-bar-rtl::before{-webkit-animation:1s ease-out progressBar-slide-complete-rtl;animation:1s ease-out progressBar-slide-complete-rtl;transform:translate3d(0,0,0)}.fast-closing.ngx-progress-bar-rtl::before{-webkit-animation:.6s ease-out progressBar-slide-complete-rtl!important;animation:.6s ease-out progressBar-slide-complete-rtl!important}@-webkit-keyframes progressBar-slide-ltr{from{transform:translate3d(-100%,0,0)}to{transform:translate3d(-5%,0,0)}}@keyframes progressBar-slide-ltr{from{transform:translate3d(-100%,0,0)}to{transform:translate3d(-5%,0,0)}}@-webkit-keyframes progressBar-slide-rtl{from{transform:translate3d(100%,0,0)}to{transform:translate3d(5%,0,0)}}@keyframes progressBar-slide-rtl{from{transform:translate3d(100%,0,0)}to{transform:translate3d(5%,0,0)}}@-webkit-keyframes progressBar-slide-complete-ltr{0%{transform:translate3d(-75%,0,0)}50%{transform:translate3d(0,0,0)}}@keyframes progressBar-slide-complete-ltr{0%{transform:translate3d(-75%,0,0)}50%{transform:translate3d(0,0,0)}}@-webkit-keyframes progressBar-slide-complete-rtl{0%{transform:translate3d(75%,0,0)}50%{transform:translate3d(0,0,0)}}@keyframes progressBar-slide-complete-rtl{0%{transform:translate3d(75%,0,0)}50%{transform:translate3d(0,0,0)}}.ngx-overlay{position:fixed;top:0;left:0;width:100%;height:100%;z-index:99998!important;background-color:rgba(40,40,40,.8);cursor:progress;display:none}.ngx-overlay.foreground-closing,.ngx-overlay.loading-foreground{display:block}.ngx-overlay.foreground-closing{opacity:0!important;transition:opacity .5s ease-out .5s}.ngx-overlay.fast-closing{transition:opacity .3s ease-out .3s!important}.ngx-overlay>.ngx-foreground-spinner{position:fixed;width:60px;height:60px;margin:0;color:#00acc1}.ngx-overlay>.ngx-loading-logo{position:fixed;margin:0;width:120px;height:120px}.ngx-overlay>.ngx-loading-text{position:fixed;margin:0;font-family:sans-serif;font-weight:400;font-size:1.2em;color:#fff}.ngx-background-spinner{position:fixed;z-index:99997!important;width:60px;height:60px;margin:0;color:#00acc1;opacity:.6;display:none}.ngx-background-spinner.background-closing,.ngx-background-spinner.loading-background{display:block}.ngx-background-spinner.background-closing{opacity:0!important;transition:opacity .7s ease-out}.ngx-background-spinner.fast-closing{transition:opacity .4s ease-out!important}.ngx-position-absolute,.ngx-position-absolute>.ngx-foreground-spinner,.ngx-position-absolute>.ngx-loading-logo,.ngx-position-absolute>.ngx-loading-text{position:absolute!important}.ngx-position-absolute.ngx-progress-bar{z-index:99996!important}.ngx-position-absolute.ngx-overlay{z-index:99995!important}.ngx-position-absolute .sk-square-jelly-box>div:nth-child(1),.ngx-position-absolute.ngx-background-spinner{z-index:99994!important}.top-left{top:30px;left:30px}.top-center{top:30px;left:50%;transform:translateX(-50%)}.top-right{top:30px;right:30px}.center-left{top:50%;left:30px;transform:translateY(-50%)}.center-center{top:50%;left:50%;transform:translate(-50%,-50%)}.center-right{top:50%;right:30px;transform:translateY(-50%)}.bottom-left{bottom:30px;left:30px}.bottom-center{bottom:30px;left:50%;transform:translateX(-50%)}.bottom-right{bottom:30px;right:30px}.sk-ball-scale-multiple,.sk-ball-scale-multiple>div{position:relative;box-sizing:border-box}.sk-ball-scale-multiple{width:100%;height:100%;font-size:0}.sk-ball-scale-multiple>div{display:inline-block;float:none;background-color:currentColor;border:0 solid currentColor;position:absolute;top:0;left:0;width:100%;height:100%;border-radius:100%;opacity:0;-webkit-animation:1s linear infinite ball-scale-multiple;animation:1s linear infinite ball-scale-multiple}.sk-ball-scale-multiple>div:nth-child(2){-webkit-animation-delay:.2s;animation-delay:.2s}.sk-ball-scale-multiple>div:nth-child(3){-webkit-animation-delay:.4s;animation-delay:.4s}@-webkit-keyframes ball-scale-multiple{0%{opacity:0;transform:scale(0)}5%{opacity:.75}100%{opacity:0;transform:scale(1)}}@keyframes ball-scale-multiple{0%{opacity:0;transform:scale(0)}5%{opacity:.75}100%{opacity:0;transform:scale(1)}}.sk-ball-spin,.sk-ball-spin>div{position:relative;box-sizing:border-box}.sk-ball-spin{width:100%;height:100%;font-size:0}.sk-ball-spin>div{display:inline-block;float:none;background-color:currentColor;border:0 solid currentColor;position:absolute;top:50%;left:50%;width:25%;height:25%;margin-top:-12.5%;margin-left:-12.5%;border-radius:100%;-webkit-animation:1s ease-in-out infinite ball-spin-clockwise;animation:1s ease-in-out infinite ball-spin-clockwise}.sk-ball-spin>div:nth-child(1){top:5%;left:50%;-webkit-animation-delay:-1.125s;animation-delay:-1.125s}.sk-ball-spin>div:nth-child(2){top:18.1801948466%;left:81.8198051534%;-webkit-animation-delay:-1.25s;animation-delay:-1.25s}.sk-ball-spin>div:nth-child(3){top:50%;left:95%;-webkit-animation-delay:-1.375s;animation-delay:-1.375s}.sk-ball-spin>div:nth-child(4){top:81.8198051534%;left:81.8198051534%;-webkit-animation-delay:-1.5s;animation-delay:-1.5s}.sk-ball-spin>div:nth-child(5){top:94.9999999966%;left:50.0000000005%;-webkit-animation-delay:-1.625s;animation-delay:-1.625s}.sk-ball-spin>div:nth-child(6){top:81.8198046966%;left:18.1801949248%;-webkit-animation-delay:-1.75s;animation-delay:-1.75s}.sk-ball-spin>div:nth-child(7){top:49.9999750815%;left:5.0000051215%;-webkit-animation-delay:-1.875s;animation-delay:-1.875s}.sk-ball-spin>div:nth-child(8){top:18.179464974%;left:18.1803700518%;-webkit-animation-delay:-2s;animation-delay:-2s}@-webkit-keyframes ball-spin{0%,100%{opacity:1;transform:scale(1)}20%{opacity:1}80%{opacity:0;transform:scale(0)}}@keyframes ball-spin{0%,100%{opacity:1;transform:scale(1)}20%{opacity:1}80%{opacity:0;transform:scale(0)}}.sk-ball-spin-clockwise,.sk-ball-spin-clockwise>div{position:relative;box-sizing:border-box}.sk-ball-spin-clockwise{width:100%;height:100%;font-size:0}.sk-ball-spin-clockwise>div{display:inline-block;float:none;background-color:currentColor;border:0 solid currentColor;position:absolute;top:50%;left:50%;width:25%;height:25%;margin-top:-12.5%;margin-left:-12.5%;border-radius:100%;-webkit-animation:1s ease-in-out infinite ball-spin-clockwise;animation:1s ease-in-out infinite ball-spin-clockwise}.sk-ball-spin-clockwise>div:nth-child(1){top:5%;left:50%;-webkit-animation-delay:-875ms;animation-delay:-875ms}.sk-ball-spin-clockwise>div:nth-child(2){top:18.1801948466%;left:81.8198051534%;-webkit-animation-delay:-.75s;animation-delay:-.75s}.sk-ball-spin-clockwise>div:nth-child(3){top:50%;left:95%;-webkit-animation-delay:-625ms;animation-delay:-625ms}.sk-ball-spin-clockwise>div:nth-child(4){top:81.8198051534%;left:81.8198051534%;-webkit-animation-delay:-.5s;animation-delay:-.5s}.sk-ball-spin-clockwise>div:nth-child(5){top:94.9999999966%;left:50.0000000005%;-webkit-animation-delay:-375ms;animation-delay:-375ms}.sk-ball-spin-clockwise>div:nth-child(6){top:81.8198046966%;left:18.1801949248%;-webkit-animation-delay:-.25s;animation-delay:-.25s}.sk-ball-spin-clockwise>div:nth-child(7){top:49.9999750815%;left:5.0000051215%;-webkit-animation-delay:-125ms;animation-delay:-125ms}.sk-ball-spin-clockwise>div:nth-child(8){top:18.179464974%;left:18.1803700518%;-webkit-animation-delay:0s;animation-delay:0s}@-webkit-keyframes ball-spin-clockwise{0%,100%{opacity:1;transform:scale(1)}20%{opacity:1}80%{opacity:0;transform:scale(0)}}@keyframes ball-spin-clockwise{0%,100%{opacity:1;transform:scale(1)}20%{opacity:1}80%{opacity:0;transform:scale(0)}}.sk-ball-spin-clockwise-fade-rotating,.sk-ball-spin-clockwise-fade-rotating>div{position:relative;box-sizing:border-box}.sk-ball-spin-clockwise-fade-rotating{font-size:0;width:100%;height:100%;-webkit-animation:6s linear infinite ball-spin-clockwise-fade-rotating-rotate;animation:6s linear infinite ball-spin-clockwise-fade-rotating-rotate}.sk-ball-spin-clockwise-fade-rotating>div{display:inline-block;float:none;background-color:currentColor;border:0 solid currentColor;position:absolute;top:50%;left:50%;width:25%;height:25%;margin-top:-12.5%;margin-left:-12.5%;border-radius:100%;-webkit-animation:1s linear infinite ball-spin-clockwise-fade-rotating;animation:1s linear infinite ball-spin-clockwise-fade-rotating}.sk-ball-spin-clockwise-fade-rotating>div:nth-child(1){top:5%;left:50%;-webkit-animation-delay:-875ms;animation-delay:-875ms}.sk-ball-spin-clockwise-fade-rotating>div:nth-child(2){top:18.1801948466%;left:81.8198051534%;-webkit-animation-delay:-.75s;animation-delay:-.75s}.sk-ball-spin-clockwise-fade-rotating>div:nth-child(3){top:50%;left:95%;-webkit-animation-delay:-625ms;animation-delay:-625ms}.sk-ball-spin-clockwise-fade-rotating>div:nth-child(4){top:81.8198051534%;left:81.8198051534%;-webkit-animation-delay:-.5s;animation-delay:-.5s}.sk-ball-spin-clockwise-fade-rotating>div:nth-child(5){top:94.9999999966%;left:50.0000000005%;-webkit-animation-delay:-375ms;animation-delay:-375ms}.sk-ball-spin-clockwise-fade-rotating>div:nth-child(6){top:81.8198046966%;left:18.1801949248%;-webkit-animation-delay:-.25s;animation-delay:-.25s}.sk-ball-spin-clockwise-fade-rotating>div:nth-child(7){top:49.9999750815%;left:5.0000051215%;-webkit-animation-delay:-125ms;animation-delay:-125ms}.sk-ball-spin-clockwise-fade-rotating>div:nth-child(8){top:18.179464974%;left:18.1803700518%;-webkit-animation-delay:0s;animation-delay:0s}@-webkit-keyframes ball-spin-clockwise-fade-rotating-rotate{100%{transform:rotate(-360deg)}}@keyframes ball-spin-clockwise-fade-rotating-rotate{100%{transform:rotate(-360deg)}}@-webkit-keyframes ball-spin-clockwise-fade-rotating{50%{opacity:.25;transform:scale(.5)}100%{opacity:1;transform:scale(1)}}@keyframes ball-spin-clockwise-fade-rotating{50%{opacity:.25;transform:scale(.5)}100%{opacity:1;transform:scale(1)}}.sk-ball-spin-fade-rotating,.sk-ball-spin-fade-rotating>div{position:relative;box-sizing:border-box}.sk-ball-spin-fade-rotating{width:100%;height:100%;font-size:0;-webkit-animation:6s linear infinite ball-spin-fade-rotate;animation:6s linear infinite ball-spin-fade-rotate}.sk-ball-spin-fade-rotating>div{display:inline-block;float:none;background-color:currentColor;border:0 solid currentColor;position:absolute;top:50%;left:50%;width:25%;height:25%;margin-top:-12.5%;margin-left:-12.5%;border-radius:100%;-webkit-animation:1s linear infinite ball-spin-fade;animation:1s linear infinite ball-spin-fade}.sk-ball-spin-fade-rotating>div:nth-child(1){top:5%;left:50%;-webkit-animation-delay:-1.125s;animation-delay:-1.125s}.sk-ball-spin-fade-rotating>div:nth-child(2){top:18.1801948466%;left:81.8198051534%;-webkit-animation-delay:-1.25s;animation-delay:-1.25s}.sk-ball-spin-fade-rotating>div:nth-child(3){top:50%;left:95%;-webkit-animation-delay:-1.375s;animation-delay:-1.375s}.sk-ball-spin-fade-rotating>div:nth-child(4){top:81.8198051534%;left:81.8198051534%;-webkit-animation-delay:-1.5s;animation-delay:-1.5s}.sk-ball-spin-fade-rotating>div:nth-child(5){top:94.9999999966%;left:50.0000000005%;-webkit-animation-delay:-1.625s;animation-delay:-1.625s}.sk-ball-spin-fade-rotating>div:nth-child(6){top:81.8198046966%;left:18.1801949248%;-webkit-animation-delay:-1.75s;animation-delay:-1.75s}.sk-ball-spin-fade-rotating>div:nth-child(7){top:49.9999750815%;left:5.0000051215%;-webkit-animation-delay:-1.875s;animation-delay:-1.875s}.sk-ball-spin-fade-rotating>div:nth-child(8){top:18.179464974%;left:18.1803700518%;-webkit-animation-delay:-2s;animation-delay:-2s}@-webkit-keyframes ball-spin-fade-rotate{100%{transform:rotate(360deg)}}@keyframes ball-spin-fade-rotate{100%{transform:rotate(360deg)}}@-webkit-keyframes ball-spin-fade{0%,100%{opacity:1;transform:scale(1)}50%{opacity:.25;transform:scale(.5)}}@keyframes ball-spin-fade{0%,100%{opacity:1;transform:scale(1)}50%{opacity:.25;transform:scale(.5)}}.sk-chasing-dots{margin:auto;width:100%;height:100%;position:absolute;text-align:center;-webkit-animation:2s linear infinite sk-chasingDots-rotate;animation:2s linear infinite sk-chasingDots-rotate}.sk-chasing-dots>div{width:60%;height:60%;display:inline-block;position:absolute;top:0;background-color:currentColor;border-radius:100%;-webkit-animation:2s ease-in-out infinite sk-chasingDots-bounce;animation:2s ease-in-out infinite sk-chasingDots-bounce}.sk-chasing-dots>div:nth-child(2){top:auto;bottom:0;-webkit-animation-delay:-1s;animation-delay:-1s}@-webkit-keyframes sk-chasingDots-rotate{100%{transform:rotate(360deg)}}@keyframes sk-chasingDots-rotate{100%{transform:rotate(360deg)}}@-webkit-keyframes sk-chasingDots-bounce{0%,100%{transform:scale(0)}50%{transform:scale(1)}}@keyframes sk-chasingDots-bounce{0%,100%{transform:scale(0)}50%{transform:scale(1)}}.sk-circle{margin:auto;width:100%;height:100%;position:relative}.sk-circle>div{width:100%;height:100%;position:absolute;left:0;top:0}.sk-circle>div::before{content:\"\";display:block;margin:0 auto;width:15%;height:15%;background-color:currentColor;border-radius:100%;-webkit-animation:1.2s ease-in-out infinite both sk-circle-bounceDelay;animation:1.2s ease-in-out infinite both sk-circle-bounceDelay}.sk-circle>div:nth-child(2){transform:rotate(30deg)}.sk-circle>div:nth-child(3){transform:rotate(60deg)}.sk-circle>div:nth-child(4){transform:rotate(90deg)}.sk-circle>div:nth-child(5){transform:rotate(120deg)}.sk-circle>div:nth-child(6){transform:rotate(150deg)}.sk-circle>div:nth-child(7){transform:rotate(180deg)}.sk-circle>div:nth-child(8){transform:rotate(210deg)}.sk-circle>div:nth-child(9){transform:rotate(240deg)}.sk-circle>div:nth-child(10){transform:rotate(270deg)}.sk-circle>div:nth-child(11){transform:rotate(300deg)}.sk-circle>div:nth-child(12){transform:rotate(330deg)}.sk-circle>div:nth-child(2)::before{-webkit-animation-delay:-1.1s;animation-delay:-1.1s}.sk-circle>div:nth-child(3)::before{-webkit-animation-delay:-1s;animation-delay:-1s}.sk-circle>div:nth-child(4)::before{-webkit-animation-delay:-.9s;animation-delay:-.9s}.sk-circle>div:nth-child(5)::before{-webkit-animation-delay:-.8s;animation-delay:-.8s}.sk-circle>div:nth-child(6)::before{-webkit-animation-delay:-.7s;animation-delay:-.7s}.sk-circle>div:nth-child(7)::before{-webkit-animation-delay:-.6s;animation-delay:-.6s}.sk-circle>div:nth-child(8)::before{-webkit-animation-delay:-.5s;animation-delay:-.5s}.sk-circle>div:nth-child(9)::before{-webkit-animation-delay:-.4s;animation-delay:-.4s}.sk-circle>div:nth-child(10)::before{-webkit-animation-delay:-.3s;animation-delay:-.3s}.sk-circle>div:nth-child(11)::before{-webkit-animation-delay:-.2s;animation-delay:-.2s}.sk-circle>div:nth-child(12)::before{-webkit-animation-delay:-.1s;animation-delay:-.1s}@-webkit-keyframes sk-circle-bounceDelay{0%,100%,80%{transform:scale(0)}40%{transform:scale(1)}}@keyframes sk-circle-bounceDelay{0%,100%,80%{transform:scale(0)}40%{transform:scale(1)}}.sk-cube-grid{width:100%;height:100%;margin:auto}.sk-cube-grid>div{width:33%;height:33%;background-color:currentColor;float:left;-webkit-animation:1.3s ease-in-out infinite sk-cubeGrid-scaleDelay;animation:1.3s ease-in-out infinite sk-cubeGrid-scaleDelay}.sk-cube-grid>div:nth-child(1){-webkit-animation-delay:.2s;animation-delay:.2s}.sk-cube-grid>div:nth-child(2){-webkit-animation-delay:.3s;animation-delay:.3s}.sk-cube-grid>div:nth-child(3){-webkit-animation-delay:.4s;animation-delay:.4s}.sk-cube-grid>div:nth-child(4){-webkit-animation-delay:.1s;animation-delay:.1s}.sk-cube-grid>div:nth-child(5){-webkit-animation-delay:.2s;animation-delay:.2s}.sk-cube-grid>div:nth-child(6){-webkit-animation-delay:.3s;animation-delay:.3s}.sk-cube-grid>div:nth-child(7){-webkit-animation-delay:0s;animation-delay:0s}.sk-cube-grid>div:nth-child(8){-webkit-animation-delay:.1s;animation-delay:.1s}.sk-cube-grid>div:nth-child(9){-webkit-animation-delay:.2s;animation-delay:.2s}@-webkit-keyframes sk-cubeGrid-scaleDelay{0%,100%,70%{transform:scale3D(1,1,1)}35%{transform:scale3D(0,0,1)}}@keyframes sk-cubeGrid-scaleDelay{0%,100%,70%{transform:scale3D(1,1,1)}35%{transform:scale3D(0,0,1)}}.sk-double-bounce{width:100%;height:100%;position:relative;margin:auto}.sk-double-bounce>div{width:100%;height:100%;border-radius:50%;background-color:currentColor;opacity:.6;position:absolute;top:0;left:0;-webkit-animation:2s ease-in-out infinite sk-doubleBounce-bounce;animation:2s ease-in-out infinite sk-doubleBounce-bounce}.sk-double-bounce>div:nth-child(2){-webkit-animation-delay:-1s;animation-delay:-1s}@-webkit-keyframes sk-doubleBounce-bounce{0%,100%{transform:scale(0)}50%{transform:scale(1)}}@keyframes sk-doubleBounce-bounce{0%,100%{transform:scale(0)}50%{transform:scale(1)}}.sk-fading-circle{margin:auto;width:100%;height:100%;position:relative}.sk-fading-circle>div{width:100%;height:100%;position:absolute;left:0;top:0}.sk-fading-circle>div::before{content:\"\";display:block;margin:0 auto;width:15%;height:15%;background-color:currentColor;border-radius:100%;-webkit-animation:1.2s ease-in-out infinite both sk-fadingCircle-FadeDelay;animation:1.2s ease-in-out infinite both sk-fadingCircle-FadeDelay}.sk-fading-circle>div:nth-child(2){transform:rotate(30deg)}.sk-fading-circle>div:nth-child(3){transform:rotate(60deg)}.sk-fading-circle>div:nth-child(4){transform:rotate(90deg)}.sk-fading-circle>div:nth-child(5){transform:rotate(120deg)}.sk-fading-circle>div:nth-child(6){transform:rotate(150deg)}.sk-fading-circle>div:nth-child(7){transform:rotate(180deg)}.sk-fading-circle>div:nth-child(8){transform:rotate(210deg)}.sk-fading-circle>div:nth-child(9){transform:rotate(240deg)}.sk-fading-circle>div:nth-child(10){transform:rotate(270deg)}.sk-fading-circle>div:nth-child(11){transform:rotate(300deg)}.sk-fading-circle>div:nth-child(12){transform:rotate(330deg)}.sk-fading-circle>div:nth-child(2)::before{-webkit-animation-delay:-1.1s;animation-delay:-1.1s}.sk-fading-circle>div:nth-child(3)::before{-webkit-animation-delay:-1s;animation-delay:-1s}.sk-fading-circle>div:nth-child(4)::before{-webkit-animation-delay:-.9s;animation-delay:-.9s}.sk-fading-circle>div:nth-child(5)::before{-webkit-animation-delay:-.8s;animation-delay:-.8s}.sk-fading-circle>div:nth-child(6)::before{-webkit-animation-delay:-.7s;animation-delay:-.7s}.sk-fading-circle>div:nth-child(7)::before{-webkit-animation-delay:-.6s;animation-delay:-.6s}.sk-fading-circle>div:nth-child(8)::before{-webkit-animation-delay:-.5s;animation-delay:-.5s}.sk-fading-circle>div:nth-child(9)::before{-webkit-animation-delay:-.4s;animation-delay:-.4s}.sk-fading-circle>div:nth-child(10)::before{-webkit-animation-delay:-.3s;animation-delay:-.3s}.sk-fading-circle>div:nth-child(11)::before{-webkit-animation-delay:-.2s;animation-delay:-.2s}.sk-fading-circle>div:nth-child(12)::before{-webkit-animation-delay:-.1s;animation-delay:-.1s}@-webkit-keyframes sk-fadingCircle-FadeDelay{0%,100%,39%{opacity:0}40%{opacity:1}}@keyframes sk-fadingCircle-FadeDelay{0%,100%,39%{opacity:0}40%{opacity:1}}.sk-folding-cube{margin:auto;width:100%;height:100%;position:relative;transform:rotateZ(45deg)}.sk-folding-cube>div{float:left;width:50%;height:50%;position:relative;transform:scale(1.1)}.sk-folding-cube>div::before{content:\"\";position:absolute;top:0;left:0;width:100%;height:100%;background-color:currentColor;-webkit-animation:2.4s linear infinite both sk-foldingCube-angle;animation:2.4s linear infinite both sk-foldingCube-angle;transform-origin:100% 100%}.sk-folding-cube>div:nth-child(2){transform:scale(1.1) rotateZ(90deg)}.sk-folding-cube>div:nth-child(3){transform:scale(1.1) rotateZ(270deg)}.sk-folding-cube>div:nth-child(4){transform:scale(1.1) rotateZ(180deg)}.sk-folding-cube>div:nth-child(2)::before{-webkit-animation-delay:.3s;animation-delay:.3s}.sk-folding-cube>div:nth-child(3)::before{-webkit-animation-delay:.9s;animation-delay:.9s}.sk-folding-cube>div:nth-child(4)::before{-webkit-animation-delay:.6s;animation-delay:.6s}@-webkit-keyframes sk-foldingCube-angle{0%,10%{transform:perspective(140px) rotateX(-180deg);opacity:0}25%,75%{transform:perspective(140px) rotateX(0);opacity:1}100%,90%{transform:perspective(140px) rotateY(180deg);opacity:0}}@keyframes sk-foldingCube-angle{0%,10%{transform:perspective(140px) rotateX(-180deg);opacity:0}25%,75%{transform:perspective(140px) rotateX(0);opacity:1}100%,90%{transform:perspective(140px) rotateY(180deg);opacity:0}}.sk-pulse{width:100%;height:100%;margin:auto}.sk-pulse>div{width:100%;height:100%;background-color:currentColor;border-radius:100%;-webkit-animation:1s ease-in-out infinite sk-pulse-scaleOut;animation:1s ease-in-out infinite sk-pulse-scaleOut}@-webkit-keyframes sk-pulse-scaleOut{0%{transform:scale(0)}100%{transform:scale(1);opacity:0}}@keyframes sk-pulse-scaleOut{0%{transform:scale(0)}100%{transform:scale(1);opacity:0}}.sk-rectangle-bounce{margin:auto;width:100%;height:100%;text-align:center;font-size:0}.sk-rectangle-bounce>div{background-color:currentColor;height:100%;width:10%;margin:0 5%;display:inline-block;-webkit-animation:1.2s ease-in-out infinite sk-rectangleBounce-stretchDelay;animation:1.2s ease-in-out infinite sk-rectangleBounce-stretchDelay}.sk-rectangle-bounce>div:nth-child(2){-webkit-animation-delay:-1.1s;animation-delay:-1.1s}.sk-rectangle-bounce>div:nth-child(3){-webkit-animation-delay:-1s;animation-delay:-1s}.sk-rectangle-bounce>div:nth-child(4){-webkit-animation-delay:-.9s;animation-delay:-.9s}.sk-rectangle-bounce>div:nth-child(5){-webkit-animation-delay:-.8s;animation-delay:-.8s}@-webkit-keyframes sk-rectangleBounce-stretchDelay{0%,100%,40%{transform:scaleY(.4)}20%{transform:scaleY(1)}}@keyframes sk-rectangleBounce-stretchDelay{0%,100%,40%{transform:scaleY(.4)}20%{transform:scaleY(1)}}.sk-rectangle-bounce-party,.sk-rectangle-bounce-party>div{position:relative;box-sizing:border-box}.sk-rectangle-bounce-party{margin:auto;width:100%;height:100%;text-align:center;font-size:0}.sk-rectangle-bounce-party>div{display:inline-block;float:none;background-color:currentColor;border:0 solid currentColor;width:10%;height:100%;margin:0 5%;border-radius:0;-webkit-animation-name:rectangle-bounce-party;animation-name:rectangle-bounce-party;-webkit-animation-iteration-count:infinite;animation-iteration-count:infinite}.sk-rectangle-bounce-party>div:nth-child(1){-webkit-animation-duration:.43s;animation-duration:.43s;-webkit-animation-delay:-.23s;animation-delay:-.23s}.sk-rectangle-bounce-party>div:nth-child(2){-webkit-animation-duration:.62s;animation-duration:.62s;-webkit-animation-delay:-.32s;animation-delay:-.32s}.sk-rectangle-bounce-party>div:nth-child(3){-webkit-animation-duration:.43s;animation-duration:.43s;-webkit-animation-delay:-.44s;animation-delay:-.44s}.sk-rectangle-bounce-party>div:nth-child(4){-webkit-animation-duration:.8s;animation-duration:.8s;-webkit-animation-delay:-.31s;animation-delay:-.31s}.sk-rectangle-bounce-party>div:nth-child(5){-webkit-animation-duration:.74s;animation-duration:.74s;-webkit-animation-delay:-.24s;animation-delay:-.24s}@-webkit-keyframes rectangle-bounce-party{0%,100%{transform:scaleY(1)}50%{transform:scaleY(.4)}}@keyframes rectangle-bounce-party{0%,100%{transform:scaleY(1)}50%{transform:scaleY(.4)}}.sk-rectangle-bounce-pulse-out,.sk-rectangle-bounce-pulse-out>div{position:relative;box-sizing:border-box}.sk-rectangle-bounce-pulse-out{margin:auto;width:100%;height:100%;text-align:center;font-size:0}.sk-rectangle-bounce-pulse-out>div{display:inline-block;float:none;background-color:currentColor;border:0 solid currentColor;width:10%;height:100%;margin:0 5%;border-radius:0;-webkit-animation:.9s cubic-bezier(.85,.25,.37,.85) infinite rectangle-bounce-pulse-out;animation:.9s cubic-bezier(.85,.25,.37,.85) infinite rectangle-bounce-pulse-out}.sk-rectangle-bounce-pulse-out>div:nth-child(3){-webkit-animation-delay:-.9s;animation-delay:-.9s}.sk-rectangle-bounce-pulse-out>div:nth-child(2),.sk-rectangle-bounce-pulse-out>div:nth-child(4){-webkit-animation-delay:-.7s;animation-delay:-.7s}.sk-rectangle-bounce-pulse-out>div:nth-child(1),.sk-rectangle-bounce-pulse-out>div:nth-child(5){-webkit-animation-delay:-.5s;animation-delay:-.5s}@-webkit-keyframes rectangle-bounce-pulse-out{0%,100%{transform:scaley(1)}50%{transform:scaley(.4)}}@keyframes rectangle-bounce-pulse-out{0%,100%{transform:scaley(1)}50%{transform:scaley(.4)}}.sk-rectangle-bounce-pulse-out-rapid,.sk-rectangle-bounce-pulse-out-rapid>div{position:relative;box-sizing:border-box}.sk-rectangle-bounce-pulse-out-rapid{margin:auto;width:100%;height:100%;text-align:center;font-size:0}.sk-rectangle-bounce-pulse-out-rapid>div{display:inline-block;float:none;background-color:currentColor;border:0 solid currentColor;width:10%;height:100%;margin:0 5%;border-radius:0;-webkit-animation:.9s cubic-bezier(.11,.49,.38,.78) infinite rectangle-bounce-pulse-out-rapid;animation:.9s cubic-bezier(.11,.49,.38,.78) infinite rectangle-bounce-pulse-out-rapid}.sk-rectangle-bounce-pulse-out-rapid>div:nth-child(3){-webkit-animation-delay:-.9s;animation-delay:-.9s}.sk-rectangle-bounce-pulse-out-rapid>div:nth-child(2),.sk-rectangle-bounce-pulse-out-rapid>div:nth-child(4){-webkit-animation-delay:-.65s;animation-delay:-.65s}.sk-rectangle-bounce-pulse-out-rapid>div:nth-child(1),.sk-rectangle-bounce-pulse-out-rapid>div:nth-child(5){-webkit-animation-delay:-.4s;animation-delay:-.4s}@-webkit-keyframes rectangle-bounce-pulse-out-rapid{0%,90%{transform:scaley(1)}80%{transform:scaley(.4)}}@keyframes rectangle-bounce-pulse-out-rapid{0%,90%{transform:scaley(1)}80%{transform:scaley(.4)}}.sk-rotating-plane{width:100%;height:100%;text-align:center;margin:auto}.sk-rotating-plane>div{width:100%;height:100%;background-color:currentColor;-webkit-animation:1.2s ease-in-out infinite sk-rotatePlane;animation:1.2s ease-in-out infinite sk-rotatePlane}@-webkit-keyframes sk-rotatePlane{0%{transform:perspective(120px) rotateX(0) rotateY(0)}50%{transform:perspective(120px) rotateX(-180.1deg) rotateY(0)}100%{transform:perspective(120px) rotateX(-180deg) rotateY(-179.9deg)}}@keyframes sk-rotatePlane{0%{transform:perspective(120px) rotateX(0) rotateY(0)}50%{transform:perspective(120px) rotateX(-180.1deg) rotateY(0)}100%{transform:perspective(120px) rotateX(-180deg) rotateY(-179.9deg)}}.sk-square-jelly-box,.sk-square-jelly-box>div{position:relative;box-sizing:border-box}.sk-square-jelly-box{width:100%;height:100%;font-size:0}.sk-square-jelly-box>div{display:inline-block;float:none;background-color:currentColor;border:0 solid currentColor}.sk-square-jelly-box>div:nth-child(1),.sk-square-jelly-box>div:nth-child(2){position:absolute;left:0;width:100%}.sk-square-jelly-box>div:nth-child(1){top:-25%;z-index:99997;height:100%;border-radius:10%;-webkit-animation:.6s linear -.1s infinite square-jelly-box-animate;animation:.6s linear -.1s infinite square-jelly-box-animate}.sk-square-jelly-box>div:nth-child(2){bottom:-9%;height:10%;background:#000;border-radius:50%;opacity:.2;-webkit-animation:.6s linear -.1s infinite square-jelly-box-shadow;animation:.6s linear -.1s infinite square-jelly-box-shadow}@-webkit-keyframes square-jelly-box-animate{17%{border-bottom-right-radius:10%}25%{transform:translateY(25%) rotate(22.5deg)}50%{border-bottom-right-radius:100%;transform:translateY(50%) scale(1,.9) rotate(45deg)}75%{transform:translateY(25%) rotate(67.5deg)}100%{transform:translateY(0) rotate(90deg)}}@keyframes square-jelly-box-animate{17%{border-bottom-right-radius:10%}25%{transform:translateY(25%) rotate(22.5deg)}50%{border-bottom-right-radius:100%;transform:translateY(50%) scale(1,.9) rotate(45deg)}75%{transform:translateY(25%) rotate(67.5deg)}100%{transform:translateY(0) rotate(90deg)}}@-webkit-keyframes square-jelly-box-shadow{50%{transform:scale(1.25,1)}}@keyframes square-jelly-box-shadow{50%{transform:scale(1.25,1)}}.sk-square-loader,.sk-square-loader>div{position:relative;box-sizing:border-box}.sk-square-loader{font-size:0;width:100%;height:100%}.sk-square-loader>div{display:inline-block;float:none;border:3px solid currentColor;width:100%;height:100%;background:0 0;border-radius:0;-webkit-animation:2s infinite square-loader;animation:2s infinite square-loader}.sk-square-loader>div:after{display:inline-block;width:100%;vertical-align:top;content:\"\";background-color:currentColor;-webkit-animation:2s ease-in infinite square-loader-inner;animation:2s ease-in infinite square-loader-inner}@-webkit-keyframes square-loader{0%{transform:rotate(0)}25%,50%{transform:rotate(180deg)}100%,75%{transform:rotate(360deg)}}@keyframes square-loader{0%{transform:rotate(0)}25%,50%{transform:rotate(180deg)}100%,75%{transform:rotate(360deg)}}@-webkit-keyframes square-loader-inner{0%,100%,25%{height:0}50%,75%{height:100%}}@keyframes square-loader-inner{0%,100%,25%{height:0}50%,75%{height:100%}}.sk-three-bounce{margin:auto;width:100%;height:100%;text-align:center}.sk-three-bounce>div{margin-top:35%;width:30%;height:30%;background-color:currentColor;border-radius:100%;display:inline-block;-webkit-animation:1.4s ease-in-out infinite both sk-threeBounce-bounceDelay;animation:1.4s ease-in-out infinite both sk-threeBounce-bounceDelay}.bottom-center>.sk-three-bounce>div,.bottom-left>.sk-three-bounce>div,.bottom-right>.sk-three-bounce>div{margin-top:70%!important}.top-center>.sk-three-bounce>div,.top-left>.sk-three-bounce>div,.top-right>.sk-three-bounce>div{margin-top:0!important}.sk-three-bounce>div:nth-child(1){-webkit-animation-delay:-.32s;animation-delay:-.32s}.sk-three-bounce>div:nth-child(2){-webkit-animation-delay:-.16s;animation-delay:-.16s}@-webkit-keyframes sk-threeBounce-bounceDelay{0%,100%,80%{transform:scale(0)}40%{transform:scale(1)}}@keyframes sk-threeBounce-bounceDelay{0%,100%,80%{transform:scale(0)}40%{transform:scale(1)}}.sk-three-strings{width:100%;height:100%}.sk-three-strings>div{position:absolute;box-sizing:border-box;width:100%;height:100%;border-radius:50%}.sk-three-strings>div:nth-child(1){left:0;top:0;-webkit-animation:1s linear infinite sk-threeStrings-rotateOne;animation:1s linear infinite sk-threeStrings-rotateOne;border-bottom:3px solid currentColor}.sk-three-strings>div:nth-child(2){right:0;top:0;-webkit-animation:1s linear infinite sk-threeStrings-rotateTwo;animation:1s linear infinite sk-threeStrings-rotateTwo;border-right:3px solid currentColor}.sk-three-strings>div:nth-child(3){right:0;bottom:0;-webkit-animation:1s linear infinite sk-threeStrings-rotateThree;animation:1s linear infinite sk-threeStrings-rotateThree;border-top:3px solid currentColor}@-webkit-keyframes sk-threeStrings-rotateOne{0%{transform:rotateX(35deg) rotateY(-45deg) rotateZ(0)}100%{transform:rotateX(35deg) rotateY(-45deg) rotateZ(360deg)}}@keyframes sk-threeStrings-rotateOne{0%{transform:rotateX(35deg) rotateY(-45deg) rotateZ(0)}100%{transform:rotateX(35deg) rotateY(-45deg) rotateZ(360deg)}}@-webkit-keyframes sk-threeStrings-rotateTwo{0%{transform:rotateX(50deg) rotateY(10deg) rotateZ(0)}100%{transform:rotateX(50deg) rotateY(10deg) rotateZ(360deg)}}@keyframes sk-threeStrings-rotateTwo{0%{transform:rotateX(50deg) rotateY(10deg) rotateZ(0)}100%{transform:rotateX(50deg) rotateY(10deg) rotateZ(360deg)}}@-webkit-keyframes sk-threeStrings-rotateThree{0%{transform:rotateX(35deg) rotateY(55deg) rotateZ(0)}100%{transform:rotateX(35deg) rotateY(55deg) rotateZ(360deg)}}@keyframes sk-threeStrings-rotateThree{0%{transform:rotateX(35deg) rotateY(55deg) rotateZ(0)}100%{transform:rotateX(35deg) rotateY(55deg) rotateZ(360deg)}}.sk-wandering-cubes{margin:auto;width:100%;height:100%;position:relative;text-align:center}.sk-wandering-cubes>div{background-color:currentColor;width:25%;height:25%;position:absolute;top:0;left:0;-webkit-animation:1.8s ease-in-out infinite sk-wanderingCubes-cubeMove;animation:1.8s ease-in-out infinite sk-wanderingCubes-cubeMove}.sk-wandering-cubes>div:nth-child(2){-webkit-animation-delay:-.9s;animation-delay:-.9s}@-webkit-keyframes sk-wanderingCubes-cubeMove{25%{transform:translateX(290%) rotate(-90deg) scale(.5)}50%{transform:translateX(290%) translateY(290%) rotate(-179deg)}50.1%{transform:translateX(290%) translateY(290%) rotate(-180deg)}75%{transform:translateX(0) translateY(290%) rotate(-270deg) scale(.5)}100%{transform:rotate(-360deg)}}@keyframes sk-wanderingCubes-cubeMove{25%{transform:translateX(290%) rotate(-90deg) scale(.5)}50%{transform:translateX(290%) translateY(290%) rotate(-179deg)}50.1%{transform:translateX(290%) translateY(290%) rotate(-180deg)}75%{transform:translateX(0) translateY(290%) rotate(-270deg) scale(.5)}100%{transform:rotate(-360deg)}}"]
            }),
            __metadata("design:paramtypes", [platformBrowser.DomSanitizer, core.ChangeDetectorRef, NgxUiLoaderService])
        ], NgxUiLoaderComponent);
        return NgxUiLoaderComponent;
    }());

    var NgxUiLoaderBlurredDirective = /** @class */ (function () {
        function NgxUiLoaderBlurredDirective(elementRef, renderer, loader) {
            this.elementRef = elementRef;
            this.renderer = renderer;
            this.loader = loader;
            this.blur = this.loader.getDefaultConfig().blur;
            this.loaderId = this.loader.getDefaultConfig().masterLoaderId;
            this.fastFadeOut = this.loader.getDefaultConfig().fastFadeOut;
        }
        /**
         * On Init event
         */
        NgxUiLoaderBlurredDirective.prototype.ngOnInit = function () {
            var _this = this;
            this.showForegroundWatcher = this.loader.showForeground$
                .pipe(operators.filter(function (showEvent) { return _this.loaderId === showEvent.loaderId; }))
                .subscribe(function (data) {
                if (data.isShow) {
                    var filterValue = "blur(" + _this.blur + "px)";
                    _this.renderer.setStyle(_this.elementRef.nativeElement, '-webkit-filter', filterValue);
                    _this.renderer.setStyle(_this.elementRef.nativeElement, 'filter', filterValue);
                }
                else {
                    setTimeout(function () {
                        if (!_this.loader.hasRunningTask(FOREGROUND, data.loaderId)) {
                            _this.renderer.setStyle(_this.elementRef.nativeElement, '-webkit-filter', 'none');
                            _this.renderer.setStyle(_this.elementRef.nativeElement, 'filter', 'none');
                        }
                    }, _this.fastFadeOut ? FAST_OVERLAY_DISAPPEAR_TIME : OVERLAY_DISAPPEAR_TIME);
                }
            });
        };
        /**
         * On destroy event
         */
        NgxUiLoaderBlurredDirective.prototype.ngOnDestroy = function () {
            if (this.showForegroundWatcher) {
                this.showForegroundWatcher.unsubscribe();
            }
        };
        NgxUiLoaderBlurredDirective.ctorParameters = function () { return [
            { type: core.ElementRef },
            { type: core.Renderer2 },
            { type: NgxUiLoaderService }
        ]; };
        __decorate([
            core.Input(),
            __metadata("design:type", Number)
        ], NgxUiLoaderBlurredDirective.prototype, "blur", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], NgxUiLoaderBlurredDirective.prototype, "loaderId", void 0);
        NgxUiLoaderBlurredDirective = __decorate([
            core.Directive({ selector: '[ngxUiLoaderBlurred]' }),
            __metadata("design:paramtypes", [core.ElementRef, core.Renderer2, NgxUiLoaderService])
        ], NgxUiLoaderBlurredDirective);
        return NgxUiLoaderBlurredDirective;
    }());

    var NgxUiLoaderModule = /** @class */ (function () {
        function NgxUiLoaderModule() {
        }
        NgxUiLoaderModule_1 = NgxUiLoaderModule;
        /**
         * forRoot
         * @returns A module with its provider dependencies
         */
        NgxUiLoaderModule.forRoot = function (ngxUiLoaderConfig) {
            return {
                ngModule: NgxUiLoaderModule_1,
                providers: [
                    {
                        provide: NGX_UI_LOADER_CONFIG_TOKEN,
                        useValue: ngxUiLoaderConfig
                    }
                ]
            };
        };
        var NgxUiLoaderModule_1;
        NgxUiLoaderModule = NgxUiLoaderModule_1 = __decorate([
            core.NgModule({
                imports: [common.CommonModule],
                declarations: [NgxUiLoaderComponent, NgxUiLoaderBlurredDirective],
                exports: [NgxUiLoaderComponent, NgxUiLoaderBlurredDirective]
            })
        ], NgxUiLoaderModule);
        return NgxUiLoaderModule;
    }());

    /**
     * Injection token for ngx-ui-loader-router configuration
     */
    var NGX_UI_LOADER_ROUTER_CONFIG_TOKEN = new core.InjectionToken('ngxUiLoaderRouterCustom.config');

    function getExcludeObj(config) {
        var strs;
        var regExps;
        if (config) {
            if (config.exclude) {
                strs = config.exclude.map(function (url) { return url.toLowerCase(); });
            }
            if (config.excludeRegexp) {
                regExps = config.excludeRegexp.map(function (regexp) { return new RegExp(regexp, 'i'); });
            }
        }
        return { strs: strs, regExps: regExps };
    }
    function isIgnored(url, excludeStrings, excludeRegexps) {
        if (excludeStrings) {
            // do not show the loader for urls in the `exclude` list
            if (excludeStrings.findIndex(function (str) { return url.toLowerCase().startsWith(str); }) !== -1) {
                return true;
            }
        }
        if (excludeRegexps) {
            // do not show the loader for urls which matches regexps in the `excludeRegexp` list
            if (excludeRegexps.findIndex(function (regexp) { return regexp.test(url); }) !== -1) {
                return true;
            }
        }
        return false;
    }

    var NgxUiLoaderRouterModule = /** @class */ (function () {
        /**
         * Constructor
         */
        function NgxUiLoaderRouterModule(parentModule, customConfig, router$1, loader) {
            var _this = this;
            if (parentModule) {
                throw new Error('[ngx-ui-loader] - NgxUiLoaderRouterModule is already loaded. It should be imported in the root `AppModule` only!');
            }
            var config = {
                loaderId: loader.getDefaultConfig().masterLoaderId,
                showForeground: true
            };
            this.exclude = getExcludeObj(customConfig);
            if (customConfig) {
                config = __assign(__assign({}, config), customConfig);
            }
            router$1.events.subscribe(function (event) {
                if (event instanceof router.NavigationStart) {
                    if (!isIgnored(event.url, _this.exclude.strs, _this.exclude.regExps)) {
                        if (config.showForeground) {
                            loader.startLoader(config.loaderId, ROUTER_LOADER_TASK_ID);
                        }
                        else {
                            loader.startBackgroundLoader(config.loaderId, ROUTER_LOADER_TASK_ID);
                        }
                    }
                }
                if (event instanceof router.NavigationEnd || event instanceof router.NavigationCancel || event instanceof router.NavigationError) {
                    if (!isIgnored(event.url, _this.exclude.strs, _this.exclude.regExps)) {
                        if (config.showForeground) {
                            loader.stopLoader(config.loaderId, ROUTER_LOADER_TASK_ID);
                        }
                        else {
                            loader.stopBackgroundLoader(config.loaderId, ROUTER_LOADER_TASK_ID);
                        }
                    }
                }
            });
        }
        NgxUiLoaderRouterModule_1 = NgxUiLoaderRouterModule;
        /**
         * forRoot
         * @returns A module with its provider dependencies
         */
        NgxUiLoaderRouterModule.forRoot = function (routerConfig) {
            return {
                ngModule: NgxUiLoaderRouterModule_1,
                providers: [
                    {
                        provide: NGX_UI_LOADER_ROUTER_CONFIG_TOKEN,
                        useValue: routerConfig
                    }
                ]
            };
        };
        var NgxUiLoaderRouterModule_1;
        NgxUiLoaderRouterModule.ctorParameters = function () { return [
            { type: NgxUiLoaderRouterModule, decorators: [{ type: core.Optional }, { type: core.SkipSelf }] },
            { type: undefined, decorators: [{ type: core.Optional }, { type: core.Inject, args: [NGX_UI_LOADER_ROUTER_CONFIG_TOKEN,] }] },
            { type: router.Router },
            { type: NgxUiLoaderService }
        ]; };
        NgxUiLoaderRouterModule = NgxUiLoaderRouterModule_1 = __decorate([
            core.NgModule({}),
            __param(0, core.Optional()), __param(0, core.SkipSelf()),
            __param(1, core.Optional()), __param(1, core.Inject(NGX_UI_LOADER_ROUTER_CONFIG_TOKEN)),
            __metadata("design:paramtypes", [NgxUiLoaderRouterModule, Object, router.Router,
                NgxUiLoaderService])
        ], NgxUiLoaderRouterModule);
        return NgxUiLoaderRouterModule;
    }());

    /**
     * Injection token for ngx-ui-loader-http configuration
     */
    var NGX_UI_LOADER_HTTP_CONFIG_TOKEN = new core.InjectionToken('ngxUiLoaderHttpCustom.config');

    var NgxUiLoaderHttpInterceptor = /** @class */ (function () {
        /**
         * Constructor
         */
        function NgxUiLoaderHttpInterceptor(customConfig, loader) {
            this.loader = loader;
            this.count = 0;
            this.config = {
                loaderId: this.loader.getDefaultConfig().masterLoaderId,
                showForeground: false
            };
            this.exclude = getExcludeObj(customConfig);
            if (customConfig) {
                this.config = __assign(__assign({}, this.config), customConfig);
            }
        }
        NgxUiLoaderHttpInterceptor.prototype.intercept = function (req, next) {
            var _this = this;
            if (isIgnored(req.url, this.exclude.strs, this.exclude.regExps)) {
                return next.handle(req);
            }
            this.count++;
            if (this.config.showForeground) {
                this.loader.startLoader(this.config.loaderId, HTTP_LOADER_TASK_ID, this.config);
            }
            else {
                this.loader.startBackgroundLoader(this.config.loaderId, HTTP_LOADER_TASK_ID, this.config);
            }
            return next.handle(req).pipe(operators.finalize(function () {
                _this.count--;
                if (_this.count === 0) {
                    if (_this.config.showForeground) {
                        _this.loader.stopLoader(_this.config.loaderId, HTTP_LOADER_TASK_ID);
                    }
                    else {
                        _this.loader.stopBackgroundLoader(_this.config.loaderId, HTTP_LOADER_TASK_ID);
                    }
                }
            }));
        };
        NgxUiLoaderHttpInterceptor.ctorParameters = function () { return [
            { type: undefined, decorators: [{ type: core.Optional }, { type: core.Inject, args: [NGX_UI_LOADER_HTTP_CONFIG_TOKEN,] }] },
            { type: NgxUiLoaderService }
        ]; };
        NgxUiLoaderHttpInterceptor = __decorate([
            core.Injectable(),
            __param(0, core.Optional()), __param(0, core.Inject(NGX_UI_LOADER_HTTP_CONFIG_TOKEN)),
            __metadata("design:paramtypes", [Object, NgxUiLoaderService])
        ], NgxUiLoaderHttpInterceptor);
        return NgxUiLoaderHttpInterceptor;
    }());

    var NgxUiLoaderHttpModule = /** @class */ (function () {
        /**
         * Constructor
         */
        function NgxUiLoaderHttpModule(parentModule) {
            if (parentModule) {
                throw new Error('[ngx-ui-loader] - NgxUiLoaderHttpModule is already loaded. It should be imported in the root `AppModule` only!');
            }
        }
        NgxUiLoaderHttpModule_1 = NgxUiLoaderHttpModule;
        /**
         * forRoot
         * @returns A module with its provider dependencies
         */
        NgxUiLoaderHttpModule.forRoot = function (httpConfig) {
            return {
                ngModule: NgxUiLoaderHttpModule_1,
                providers: [
                    {
                        provide: NGX_UI_LOADER_HTTP_CONFIG_TOKEN,
                        useValue: httpConfig
                    }
                ]
            };
        };
        var NgxUiLoaderHttpModule_1;
        NgxUiLoaderHttpModule.ctorParameters = function () { return [
            { type: NgxUiLoaderHttpModule, decorators: [{ type: core.Optional }, { type: core.SkipSelf }] }
        ]; };
        NgxUiLoaderHttpModule = NgxUiLoaderHttpModule_1 = __decorate([
            core.NgModule({
                providers: [
                    {
                        provide: http.HTTP_INTERCEPTORS,
                        useClass: NgxUiLoaderHttpInterceptor,
                        multi: true
                    }
                ]
            }),
            __param(0, core.Optional()), __param(0, core.SkipSelf()),
            __metadata("design:paramtypes", [NgxUiLoaderHttpModule])
        ], NgxUiLoaderHttpModule);
        return NgxUiLoaderHttpModule;
    }());

    exports.NgxUiLoaderHttpModule = NgxUiLoaderHttpModule;
    exports.NgxUiLoaderModule = NgxUiLoaderModule;
    exports.NgxUiLoaderRouterModule = NgxUiLoaderRouterModule;
    exports.NgxUiLoaderService = NgxUiLoaderService;
    exports.ɵa = NGX_UI_LOADER_CONFIG_TOKEN;
    exports.ɵb = NgxUiLoaderComponent;
    exports.ɵc = NgxUiLoaderBlurredDirective;
    exports.ɵd = NGX_UI_LOADER_ROUTER_CONFIG_TOKEN;
    exports.ɵe = NgxUiLoaderHttpInterceptor;
    exports.ɵf = NGX_UI_LOADER_HTTP_CONFIG_TOKEN;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
