// ngx-ui-loader/esm5/lib/core/ngx-ui-loader.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NgxUiLoaderComponent } from './ngx-ui-loader.component';
import { NgxUiLoaderBlurredDirective } from './ngx-ui-loader-blurred.directive';
import { NGX_UI_LOADER_CONFIG_TOKEN } from './ngx-ui-loader-config.token';
var NgxUiLoaderModule = /** @class */ (function () {
    function NgxUiLoaderModule() {
    }
    NgxUiLoaderModule_1 = NgxUiLoaderModule;
    /**
     * forRoot
     * @returns A module with its provider dependencies
     */
    NgxUiLoaderModule.forRoot = function (ngxUiLoaderConfig) {
        return {
            ngModule: NgxUiLoaderModule_1,
            providers: [
                {
                    provide: NGX_UI_LOADER_CONFIG_TOKEN,
                    useValue: ngxUiLoaderConfig
                }
            ]
        };
    };
    var NgxUiLoaderModule_1;
    NgxUiLoaderModule = NgxUiLoaderModule_1 = __decorate([
        NgModule({
            imports: [CommonModule],
            declarations: [NgxUiLoaderComponent, NgxUiLoaderBlurredDirective],
            exports: [NgxUiLoaderComponent, NgxUiLoaderBlurredDirective]
        })
    ], NgxUiLoaderModule);
    return NgxUiLoaderModule;
}());
export { NgxUiLoaderModule };
