// ngx-logger/fesm2015/ngx-logger-testing.js
import { LoggerConfig, LoggerModule, NGXLogger, NGXLoggerHttpService, CustomNGXLoggerService, NGXMapperService } from 'ngx-logger';
import { of } from 'rxjs';
import { NgModule } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class NGXLoggerMock {
    constructor() {
    }
    /**
     * @param {?} message
     * @param {...?} additional
     * @return {?}
     */
    trace(message, ...additional) {
    }
    /**
     * @param {?} message
     * @param {...?} additional
     * @return {?}
     */
    debug(message, ...additional) {
    }
    /**
     * @param {?} message
     * @param {...?} additional
     * @return {?}
     */
    info(message, ...additional) {
    }
    /**
     * @param {?} message
     * @param {...?} additional
     * @return {?}
     */
    log(message, ...additional) {
    }
    /**
     * @param {?} message
     * @param {...?} additional
     * @return {?}
     */
    warn(message, ...additional) {
    }
    /**
     * @param {?} message
     * @param {...?} additional
     * @return {?}
     */
    error(message, ...additional) {
    }
    /**
     * @param {?} message
     * @param {...?} additional
     * @return {?}
     */
    fatal(message, ...additional) {
    }
    /**
     * @param {?} config
     * @return {?}
     */
    updateConfig(config) {
    }
    /**
     * @param {?} headers
     * @return {?}
     */
    setCustomHttpHeaders(headers) {
    }
    /**
     * @param {?} params
     * @return {?}
     */
    setCustomParams(params) {
    }
    /**
     * @param {?} monitor
     * @return {?}
     */
    registerMonitor(monitor) {
    }
    /**
     * @param {?} withCredentials
     * @return {?}
     */
    setWithCredentialsOptionValue(withCredentials) {
    }
    /**
     * @return {?}
     */
    getConfigSnapshot() {
        return new LoggerConfig();
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * CustomNGXLoggerServiceMock is a mock for CustomNGXLoggerService
 */
class CustomNGXLoggerServiceMock {
    constructor() {
    }
    /**
     * @return {?}
     */
    create() {
        // you can inject your own httpService or use the default,
        return new NGXLoggerMock();
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class NGXLoggerHttpServiceMock {
    constructor() {
    }
    /**
     * @param {?} url
     * @param {?} message
     * @param {?} additional
     * @param {?} timestamp
     * @param {?} logLevel
     * @return {?}
     */
    logOnServer(url, message, additional, timestamp, logLevel) {
        return of({});
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class NGXMapperServiceMock {
    constructor() {
    }
    /**
     * @return {?}
     */
    getCallerDetails() {
        return of({ fileName: 'test.ts', lineNumber: 0, columnNumber: 0 });
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class LoggerTestingModule {
}
LoggerTestingModule.decorators = [
    { type: NgModule, args: [{
                imports: [LoggerModule],
                providers: [
                    { provide: NGXLogger, useClass: NGXLoggerMock },
                    { provide: NGXLoggerHttpService, useClass: NGXLoggerHttpServiceMock },
                    { provide: CustomNGXLoggerService, useClass: CustomNGXLoggerServiceMock },
                    { provide: NGXMapperService, useClass: NGXMapperServiceMock }
                ]
            },] }
];

export { CustomNGXLoggerServiceMock, LoggerTestingModule, NGXLoggerHttpServiceMock, NGXLoggerMock, NGXMapperServiceMock };
